### DISCUSSION TITLE
Question: Jobs gets stuck in running state when it throws an unhandled exception

### DISCUSSION THREAD

**Discussion by Sicos1977** [quartznet/quartznet#908] [General] [answered]
Hi,

I've got a question, now and then I sometimes have a job that just gets stuck in a running state in the scheduler without doing anything. Since I upgraded to QuartzNet 3.1 it seems to be logging some more information.

I get these errors from a joblistener that I have added.

```
Unable to notify JobListener(s) of Job that was executed: (error will be ignored). trigger= SystemJob.6@Controleer documenten queue job= SystemJob.6@Controleer documenten queue
JobListener 'DocumentServices joblistener' threw exception: De connectie string 'EConversie' is niet gevonden in het configuratie bestand 'e:/Apps/DocumentServices/DocumentServices.Server.BusinessLogic.DLL.config', fout: Object reference not set to an instance of an object.
```

`Job SystemJob.6@Controleer documenten queue threw an exception. Job threw an unhandled exception.`

I now know in this case what the error is, but the problem that I have is that this jobs just gets stuck in the scheduler doing nothing. Is there a way to detect this or change the code so that it doesn't get stuck when an unhandled exception is thrown?

Also the weird problem is that whatever job is started after the job that is stuck then this second job also gets stuck (doing nothing).

![image](https://user-images.githubusercontent.com/6692947/89018394-1e5a4980-d31c-11ea-93fb-a299b8c27f14.png)

The job with the red arrow is the job that throws the unhandled exception. The other job is the one that also gets stuck.

**Comment by lahma** [comment, marked as answer]
The general rule for listeners is that they should never throw. This is easier to require on Java side where API/interface doesn't allow throwing an exception. This would probably require you to do/enable logging on more detailed level to find out what the actual problem is. Quartz will be in undefined state as it's not sure whether job was completed / logic handled correctly.

**Comment by lahma** [comment]
I've clarified this [in the documentation](https://www.quartz-scheduler.net/documentation/quartz-3.x/tutorial/trigger-and-job-listeners.html) as this is an easy trap to fall into.

**Comment by Sicos1977** [comment]
I already had try ... catch code added but I added some more to see if this will catch any leftover exceptions.

**Comment by Sicos1977** [comment]
In that case I think my problem sometimes occures overhere. I'm handling the information overthere that is returned from my jobs. I can imagine that the database where the information is stored sometimes does not respond thus throwing an error. So I also wrapped that in a try and catch.... lets hope that this solves the problem

```c#
public Task JobWasExecuted(IJobExecutionContext context, JobExecutionException jobException, CancellationToken cancellationToken)
{
    try
    {
        if (context.Result is JobResult jobResult)
        {
            switch (jobResult.Status)
            {
                case JobResultStatus.SuccesWithNoResult:
                    if (JobWithNoResultLogging)
                        _jobLogger.WriteToJobLogs(JobStatus.Completed, context);
                    break;

                case JobResultStatus.SuccessWithResult:
                    _jobLogger.WriteToJobLogs(JobStatus.Completed, context);
                    break;

                case JobResultStatus.Error:
                    _jobLogger.WriteToJobLogs(JobStatus.Error, context);
                    break;
            }
        }

        if (jobException != null)
        {
            _jobLogger.WriteToJobLogs(JobStatus.Error, context);
            _jobLogger.WriteToErrorLogs(jobException);
        }

        // Als er geen actieve triggers meer zijn disable dan de job
        if (!context.NextFireTimeUtc.HasValue)
        {
            var parts = context.Trigger.Key.Name.Split('@');
            if (parts.Length == 2)
            {
                try
                {
                    var jobId = int.Parse(parts[0]);
                    // Jobs met id = 0 zijn vanuit de beheertool aangemaakte jobs, deze bestaan niet in de database
                    // en kunnen we dus ook niet opzoeken
                    if (jobId != 0)
                    {
                        var job = Jobs.Instance.SelectForId(jobId);
                        job.Enabled = false;
                        Jobs.Instance.InsertUpdate(job);
                        _jobLogger.WriteToSystemLogs(
                            $"Job '{parts[1]}' uitgeschakeld omdat deze geen actieve triggers meer heeft");
                    }
                }
                catch (Exception exception)
                {
                    _jobLogger.WriteToErrorLogs(exception);
                }
            }
        }
    }
    catch (Exception exception)
    {
        _jobLogger.WriteToErrorLogs(exception);
    }

    return Task.CompletedTask;
}
```
